src/main/java/io/reactivex/internal/operators/maybe/MaybeFromObservable.java

```diff
@@ -10,47 +10,107 @@
  * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
  * the License for the specific language governing permissions and limitations under the License.
  */
+
 package io.reactivex.internal.operators.maybe;
 
-import io.reactivex.Observable;
-import io.reactivex.Observer;
-import io.reactivex.Maybe;
-import io.reactivex.MaybeObserver;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.reactivex.*;
 import io.reactivex.disposables.Disposable;
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.fuseable.HasUpstreamObservableSource;
+import io.reactivex.plugins.RxJavaPlugins;
 
-public final class MaybeFromObservable<T> extends Maybe<T> {
-    private final Observable<T> upstream;
+/**
+ * Wraps a publisher, expects it to emit 0 or 1 signal or an error.
+ *
+ * @param <T> the value type
+ */
+public final class MaybeFromObservable<T> extends Maybe<T> implements HasUpstreamObservableSource<T> {
 
-    public MaybeFromObservable(Observable<T> upstream) {
-        this.upstream = upstream;
+    final ObservableSource<T> source;
+    
+    public MaybeFromObservable(ObservableSource<T> source) {
+        this.source = source;
     }
-
+    
+    @Override
+    public ObservableSource<T> source() {
+        return source;
+    }
+    
     @Override
-    protected void subscribeActual(final MaybeObserver<? super T> s) {
-        upstream.subscribe(new Observer<T>() {
-            T last;
-            @Override
-            public void onSubscribe(Disposable d) {
-                s.onSubscribe(d);
+    protected void subscribeActual(MaybeObserver<? super T> observer) {
+        source.subscribe(new FromPublisherToMaybeObserver<T>(observer));
+    }
+    
+    static final class FromPublisherToMaybeObserver<T> 
```

Comment:
not that it really matters, but FromPublisherToMaybeObserver -> FromObservableToMaybeObserver